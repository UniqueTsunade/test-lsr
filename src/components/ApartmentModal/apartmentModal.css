:root {
  --card-padding: 1rem;
  --card-radius: 8px;
  --card-gap: 0.75rem;
  --card-border: 1px solid #ddd;
  --card-bg: #fff;
  --card-shadow: 0 2px 4px rgba(0, 0, 0, 0.05);
  --main-text: 1.6rem;
  --label-text: 1.1rem;
}

.apartment-modal__overlay {
  position: fixed;
  top: 0;
  left: 0;
  width: 100vw;
  height: 100vh;
  background: rgba(0, 0, 0, 0.8);
  display: flex;
  align-items: center;
  justify-content: center;
  z-index: 9999;
  cursor: pointer;
}

.apartment-modal__content {
  position: relative;
  background: #fff;
  padding: 2rem;
  border-radius: 8px;
  max-width: 600px;
  width: 100%;
  box-shadow: 0 2px 10px rgba(0, 0, 0, 0.2);
  cursor: auto;
}

.apartment-modal__description {
  margin-top: 1.4rem;
  font-size: var(--label-text);
}
