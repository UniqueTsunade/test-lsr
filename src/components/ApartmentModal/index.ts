import ApartmentModal from './ApartmentModal'

export default ApartmentModal
