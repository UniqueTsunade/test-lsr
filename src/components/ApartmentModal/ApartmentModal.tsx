import './apartmentModal.css'
import React, { useEffect, useRef } from 'react'
import ReactDOM from 'react-dom'
import ApartmentCardPreview from '../ApartmentCardPreview'
import FavouriteButton from '../FavouriteButton'
import { Apartment } from '../../types'

type ApartmentModalProps = {
  apartment: Apartment
  isFavourite: boolean
  onToggleFavourite: () => void
  onClose: () => void
}

const ApartmentModal: React.FC<ApartmentModalProps> = ({
  apartment,
  isFavourite,
  onToggleFavourite,
  onClose,
}) => {
  const modalRef = useRef<HTMLDivElement>(null)

  useEffect(() => {
    modalRef.current?.focus()
  }, [])

  const handleKeyDown = (e: React.KeyboardEvent) => {
    if (['Enter', ' ', 'Escape'].includes(e.key)) {
      e.preventDefault()
      e.stopPropagation()
      onClose()
    }
  }

  const modalRoot = document.getElementById('modal-root')
  if (!modalRoot) return null

  return ReactDOM.createPortal(
    <div
      ref={modalRef}
      className="apartment-modal__overlay"
      role="button"
      onKeyDown={handleKeyDown}
      tabIndex={0}
      onClick={onClose}
      data-testid="modal-overlay"
    >
      {/* eslint-disable-next-line jsx-a11y/no-static-element-interactions, jsx-a11y/click-events-have-key-events */}
      <div
        className="apartment-modal__content"
        data-testid="modal-content"
        onClick={(e) => e.stopPropagation()}
      >
        <ApartmentCardPreview {...apartment} />
        <div className="apartment-modal__description">
          {apartment.description}
        </div>
        <div className="apartment-modal__favourite">
          <FavouriteButton
            isFavourite={isFavourite}
            onToggle={onToggleFavourite}
          />
        </div>
      </div>
    </div>,
    modalRoot
  )
}

export default ApartmentModal
