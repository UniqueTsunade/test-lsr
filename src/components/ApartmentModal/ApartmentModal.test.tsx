import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import ApartmentModal from './ApartmentModal'

const mockApartment = {
  building: '130',
  floor: '32/2',
  type: '3-комнатная квартира',
  price: 32428252,
  image: '/test-image.jpg',
  description: 'Пример описания квартиры для тестирования.',
}

const renderModal = (
  onClose = jest.fn(),
  isFavourite = false,
  onToggleFavourite = jest.fn()
) =>
  render(
    <ApartmentModal
      apartment={mockApartment}
      onClose={onClose}
      isFavourite={isFavourite}
      onToggleFavourite={onToggleFavourite}
    />
  )

test('вызывает onClose при нажатии Escape', async () => {
  const user = userEvent.setup()
  const onClose = jest.fn()

  renderModal(onClose)
  await user.keyboard('{Escape}')

  expect(onClose).toHaveBeenCalled()
})

test('вызывает onClose при клике по затемнению', async () => {
  const user = userEvent.setup()
  const onClose = jest.fn()

  renderModal(onClose)

  const overlay = screen.getByTestId('modal-overlay')
  await user.click(overlay)

  expect(onClose).toHaveBeenCalled()
})

test('не вызывает onClose при клике на контент модалки', async () => {
  const user = userEvent.setup()
  const onClose = jest.fn()

  renderModal(onClose)

  const modalContent = screen.getByTestId('modal-content')
  await user.click(modalContent)

  expect(onClose).not.toHaveBeenCalled()
})

test('отображает кнопку избранного', () => {
  renderModal()

  expect(screen.getByLabelText(/в избранное/i)).toBeInTheDocument()
})
test('рендерится корректно (snapshot)', () => {
  const { asFragment } = renderModal()
  expect(asFragment()).toMatchSnapshot()
})
