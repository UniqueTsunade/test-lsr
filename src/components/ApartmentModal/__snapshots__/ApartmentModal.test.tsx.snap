// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`рендерится корректно (snapshot) 1`] = `<DocumentFragment />`;
